cpp/src/strings/char_types/char_cases.cu

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <vector>
+#include <map>
+#include <algorithm>
+
+#include <cudf/cudf.h>
+#include <cudf/utilities/error.hpp>
+
+//
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+namespace {
+
+  struct special_case_mapping_in {
+    uint16_t num_upper_chars;
+    uint16_t upper[3];
+    uint16_t num_lower_chars;
+    uint16_t lower[3];
+  };
+  constexpr special_case_mapping_in codepoint_mapping_in[] = {
```

Comment:
Use `std::array`.

Reply:
I talked with Jake about this previously and the issue was that if you use std::array you have to hardcode the size which is pretty clunky for an array of this size.

Reply:
Hmmm, isn't the size of unicode alphabets well defined? At least use algorithms as suggested (they don't require std::array).

At a minimum, make your `sizeof(array) / sizeof(type)` calculation robust to changes by defining a type alias so that you don't risk changing the type of the array and not the size calculation.

Reply:
See comment below.